tensorflow/python/client/timeline.py

```diff
@@ -607,7 +617,48 @@ def _show_memory_counters(self):
                                         total_bytes)
     self._allocator_maximums = alloc_maxes
 
-  def analyze_step_stats(self, show_dataflow=True, show_memory=True):
+  def _preprocess_op_time(self, op_time):
+    """Modified the start and end of ops in step stats"""
+    if op_time == "schedule":
+      self._step_stats = self._origin_step_stats
+      return
+    self._step_stats = copy.deepcopy(self._origin_step_stats)
+    # Separate job task and gpu tracer stream
+    stream_all_stats = []
+    job_stats = []
+    for stats in self._step_stats.dev_stats:
+      if "/stream:all" in stats.device:
+        stream_all_stats.append(stats)
+      elif "/job" in stats.device:
+        job_stats.append(stats)
+
+    # Record the start time of the first kernel and the end time of
+    # the last kernel for all ops.
+    op_gpu_start = {}
+    op_gpu_end = {}
+    for stats in stream_all_stats:
+      for kernel in stats.node_stats:
+        name, _ = self._parse_kernel_label(kernel.timeline_label, kernel.node_name)
+        start = kernel.all_start_micros
+        end = kernel.all_start_micros + kernel.all_end_rel_micros
+        if name in op_gpu_start:
+          op_gpu_start[name] = min(op_gpu_start[name], start)
+          op_gpu_end[name] = max(op_gpu_end[name], end)
+        else:
+          op_gpu_start[name] = start
+          op_gpu_end[name] = end
+
+    # Update the start and end time of each op according to the op_timement
+    for stats in job_stats:
+      for op in stats.node_stats:
+        if op.node_name in op_gpu_start:
+          end = max(op_gpu_end[op.node_name], op.all_start_micros + op.all_end_rel_micros)
+          if op_time == "gpu":
+            op.all_start_micros = op_gpu_start[op.node_name]
+          op.all_end_rel_micros = end - op.all_start_micros
+
+  def analyze_step_stats(self, show_dataflow=True, show_memory=True, op_time="schedule"):
+    self._preprocess_op_time(op_time)
     self._allocate_pids()
```

Comment:
Can you add a docstring for this function as well?

Reply:
Added.

Reply:
supernit: "Analyze the step stats and format it into Chrome Trace Format" => "Analyze the step stats and format it into Chrome Trace Format."
One-line docstring summary should end with ".", ".)", "?", or "!" , otherwise it triggers a lint error internally.
Can you also run pylint to check your Python changes?
https://www.tensorflow.org/community/contribute/code_style

Reply:
Thank you for pointing out the lint problem. I've added the period in the comment and made all lines less than 80 characters. Right now the remaining pylint alerts have nothing to do with PR. The pylint for `timeline.py` now is 
```
************* Module tensorflow.python.client.timeline
tensorflow/python/client/timeline.py:41:2: W0107: Unnecessary pass statement (unnecessary-pass)
tensorflow/python/client/timeline.py:52:2: W0107: Unnecessary pass statement (unnecessary-pass)
tensorflow/python/client/timeline.py:55:0: R0205: Class '_ChromeTraceFormatter' inherits from object, can be safely removed from bases in python3 (useless-object-inheritance)
tensorflow/python/client/timeline.py:259:4: R1705: Unnecessary "else" after "return" (no-else-return)
tensorflow/python/client/timeline.py:265:0: R0205: Class '_TensorTracker' inherits from object, can be safely removed from bases in python3 (useless-object-inheritance)
tensorflow/python/client/timeline.py:346:0: R0205: Class 'Timeline' inherits from object, can be safely removed from bases in python3 (useless-object-inheritance)
tensorflow/python/client/timeline.py:528:4: R1702: Too many nested blocks (6/5) (too-many-nested-blocks)

------------------------------------------------------------------
Your code has been rated at 9.78/10 (previous run: 9.78/10, +0.00)
```